src/shake.jl

```diff
@@ -0,0 +1,128 @@
+abstract type SHAKE <: SHA3_CTX end 
+# note, that field property used has differend uses, depending on T<:SHAKE or T<:SHA3_CTX
+mutable struct SHAKE_128_CTX <: SHAKE
+    state::Array{UInt64,1}
+    bytecount::UInt128
+    buffer::Array{UInt8,1}
+    bc::Array{UInt64,1}
+    used::Bool
+end
+mutable struct SHAKE_256_CTX <: SHAKE
+    state::Array{UInt64,1}
+    bytecount::UInt128
+    buffer::Array{UInt8,1}
+    bc::Array{UInt64,1}
+    used::Bool
+end
+
+digestlen(::Type{SHAKE_128_CTX}) = 16
+digestlen(::Type{SHAKE_256_CTX}) = 32
+blocklen(::Type{SHAKE_128_CTX}) = UInt64(25*8 - 2*digestlen(SHAKE_128_CTX))
+blocklen(::Type{SHAKE_256_CTX}) = UInt64(25*8 - 2*digestlen(SHAKE_256_CTX))
+buffer_pointer(ctx::T) where {T<:SHAKE} = Ptr{state_type(T)}(pointer(ctx.buffer))
+
+# construct an empty SHA context
+SHAKE_128_CTX() = SHAKE_128_CTX(zeros(UInt64, 25), 0, zeros(UInt8, blocklen(SHAKE_128_CTX)), Vector{UInt64}(undef, 5), false)
+SHAKE_256_CTX() = SHAKE_256_CTX(zeros(UInt64, 25), 0, zeros(UInt8, blocklen(SHAKE_256_CTX)), Vector{UInt64}(undef, 5), false)
+
+function transform!(context::T) where {T<:SHAKE}
+    # First, update state with buffer
+    pbuf = Ptr{eltype(context.state)}(pointer(context.buffer))
+    # after SHAKE_256_MAX_READ (digestlen) is reached, simply work with context.state[idx]
+    if !context.used 
+        for idx in 1:div(blocklen(T),8)
+            context.state[idx] = context.state[idx] ⊻ unsafe_load(pbuf, idx)
+        end
+    end 
+    bc = context.bc
+    state = context.state
+    # We always assume 24 rounds
+    @inbounds for round in 0:23
+        # Theta function
+        for i in 1:5
+            bc[i] = state[i] ⊻ state[i + 5] ⊻ state[i + 10] ⊻ state[i + 15] ⊻ state[i + 20]
+        end
+        for i in 0:4
+            temp = bc[rem(i + 4, 5) + 1] ⊻ L64(1, bc[rem(i + 1, 5) + 1])
+            j = 0
+            while j <= 20
+                state[Int(i + j + 1)] = state[i + j + 1] ⊻ temp
+                j += 5
+            end
```

Comment:
Perhaps this should be a `for` loop, just to make this kind of stuff easier for the compiler to vectorize?

```suggestion
            for j in 0:5:20
                state[Int(i + j + 1)] = state[i + j + 1] ⊻ temp
            end
```

Reply:
yes, i implemented this all together with the docstrings in my last commit.